File: LeetCode/src/main/java/com/jack/stack/LeetCode20.java
package com.jack.stack;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

public class LeetCode20 {
    public static void main(String[] args) {
        
    }

    static class Solution {
        
        public boolean isValid(String s) {
            Map<Character, Character> map = new HashMap<>();
            map.put('(',')');
            map.put('[',']');
            map.put('{','}');

            Deque<Character> stack = new ArrayDeque<>();

            for(char c : s.toCharArray()){

                if (map.containsKey(c)){
                    stack.push(c);
                    continue;
                }

                if (!map.isEmpty() && map.get(stack.peek())==c){
                    stack.pop();
                    continue;
                }

                return false;
            }

            return stack.isEmpty();
        }
    }
}

File: LeetCode/src/main/java/com/jack/stack/LeetCode20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        d = {'()', '[]', '{}'}
        
        for c in s:
            if c in '({[':
                stk.append(c)
            elif not stk or stk.pop() + c not in d:
                return False
        return not stk



if __name__ == "__main__":
    str = "()[]{}(("
    res = Solution().isValid(str)
    print(res)

